### PR TITLE
Drop support for Python 3.9

### PR DESCRIPTION
**Pull request by binaryDiv**
This PR removes support for Python 3.9, raising the minimum Python version to 3.10, in order to allow for more modern typing (union types with `X | Y`), and modernizes the code base.

Additionally, this allowed us to remove a rather hacky workaround that was needed for Python < 3.10, where dataclasses did not support `kw_only` yet. This also removes the need for `coverage-conditional-plugin`.

I also updated the testing dependencies.

This qualifies as a **breaking change** in terms of Python version compatibility. There are no breaking API changes, however.

### CODE REVIEW

File: tests/dataclasses/validataclass_test.py [hunk 1]
```diff
@@ -62,9 +59,9 @@ class UnitTestValidatorDataclass:
 
         # Check names and types of all fields
         assert list(fields.keys()) == ['foo', 'bar', 'baz']
-        assert fields['foo'].type is int
-        assert fields['bar'].type is int
-        assert fields['baz'].type is Optional[str]
+        assert fields['foo'].type == int
```
Comment (the-infinity) [inline tests/dataclasses/validataclass_test.py:62]: Just out of interest: why did you change this? :)

File: docs/05-dataclasses.md
Comment (JGaukrogers) [inline docs/05-dataclasses.md:1]: In the example code of this file, I can still see that `from typing import Optional` is still used, for example in the line 460. Is this wanted, or do we want to start using only `TypeName|None`?

**Comment by JGaukrogers** [comment]
I see no changes in the file `validataclass_mixin_test.py`, but there is the method `test_create_with_defaults_invalid_required_fields`, which contains the comment `# The exact exception message changed between Python versions 3.9 and 3.10`. Now I am wondering if there needs to be adapted something, or if we can leave it as it is.

File: tests/dataclasses/validataclass_test.py (hunk 1 above)
Comment (binaryDiv) [reply tests/dataclasses/validataclass_test.py:62]: You mean changing the `is` to `==`?

One reason is that mypy was complaining about it (see the `# mypy: no-strict-equality` in the file header that I have removed now).

But also, `==` is sufficient anyway and it makes working with `Union` and other constructs easier. For example, `int | str` and `typing.Union[int, str]` are equal (in the sense of `==`), but technically they aren't the same thing (in the sense of `is`) because they're represented by different classes:

```py
>>> from typing import Union
>>> int | str == Union[int, str]
True
>>> int | str is Union[int, str]
False
>>> type(Union[int, str])
<class 'typing._UnionGenericAlias'>
>>> type(int | str)
<class 'types.UnionType'>
```

Comment (the-infinity) [reply tests/dataclasses/validataclass_test.py:62]: Ui. Interesting :)

File: docs/05-dataclasses.md
Comment (binaryDiv) [reply docs/05-dataclasses.md:1]: I intentionally kept it that way for this particular example, because it feels a bit more consistent and intuitive to me when there's `Optional[int]` and `OptionalUnset[int]` in the example, rather than `int | None` and `OptionalUnset[int]`.

Generally, I think using `T | None` as the default makes sense (it's shorter, doesn't require any imports, is more consistent with other type unions, etc.), but as far as I know `Optional` is not deprecated (yet?), so I think it's valid to use it in places like here when it somehow improves readability or consistency.

**Comment by binaryDiv** [comment]
@JGaukrogers 

> I see no changes in the file `validataclass_mixin_test.py`, but there is the method `test_create_with_defaults_invalid_required_fields`, which contains the comment `# The exact exception message changed between Python versions 3.9 and 3.10`. Now I am wondering if there needs to be adapted something, or if we can leave it as it is.

Well spotted! :D

I was thinking about changing that too, but I don't think it's really worth changing. For one thing, just partially matching the error message is sufficient I think, and also it's a test for a deprecated function that will be removed soon, anyway.

(Actually, I think it's probably a good time to remove it in the upcoming release, it's been deprecated for a while and I don't think it's still used anywhere in our projects... But that's for another PR ^^)
